### ISSUE TITLE
Functions hashcat_session_init/folder_config_init do not respect install_folder or shared_folder

### ISSUE THREAD

**Issue by process0** [hashcat/hashcat#3911]
**Describe the bug**
The functions `hashcat_session_init` and consequently `folder_config_init` do not respect the `install_folder` or `shared_folder` arguments.  As a caller of this function, I would expect the arguments to be honored and not discarded due to implicit folder decision logic. This comment indicates others also experience this issue https://github.com/hashcat/hashcat/issues/20#issuecomment-1465196363. My platform is POSIX. 

The logic, as I understand it, compares the provided `install_folder` argument with the `install_dir` (directory of process executable).  If they are the same, then the `profile/cache/session_dir` will be set from the `home_dir` and `shared_dir` will be set to `shared_folder`. However if `install_folder` is not the same as `install_dir`, then all the directories are set to `install_dir` (directory of process executable).  

The `folder_config_init` function abbreviated:
```c
int folder_config_init (hashcat_ctx_t *hashcat_ctx, MAYBE_UNUSED const char *install_folder, MAYBE_UNUSED const char *shared_folder)
{
...
  char *exec_path = (char *) hcmalloc (exec_path_sz);

  const int rc = get_exec_path (exec_path, exec_path_sz);
...
  #if defined (_POSIX)

  static const char SLASH[] = "/";

  if (install_folder == NULL) install_folder = SLASH; // makes library use easier

  char *resolved_install_folder = realpath (install_folder, NULL);
  char *resolved_exec_path      = realpath (exec_path, NULL);

  if (resolved_install_folder == NULL) resolved_install_folder = hcstrdup (SLASH);
...

  char *install_dir = (char *) hcmalloc (HCBUFSIZ_TINY);

  get_install_dir (install_dir, resolved_exec_path);

  char *profile_dir = NULL;
  char *cache_dir   = NULL;
  char *session_dir = NULL;
  char *shared_dir  = NULL;

  if (strcmp (install_dir, resolved_install_folder) == 0)
  {
    struct passwd pw;
    struct passwd *pwp = NULL;

    char buf[HCBUFSIZ_TINY];

    memset (buf, 0, sizeof (buf));
    memset (&pw, 0, sizeof (pw));

    getpwuid_r (getuid (), &pw, buf, HCBUFSIZ_TINY, &pwp);

    const char *home_dir = pwp->pw_dir;

    profile_dir = (char *) hcmalloc (HCBUFSIZ_TINY);
    cache_dir   = (char *) hcmalloc (HCBUFSIZ_TINY);
    session_dir = (char *) hcmalloc (HCBUFSIZ_TINY);

    get_profile_dir (profile_dir, home_dir);
    get_cache_dir   (cache_dir,   home_dir);
    get_session_dir (session_dir, profile_dir);

    shared_dir = hcstrdup (shared_folder);

    hc_mkdir_rec (profile_dir, 0700);
    hc_mkdir_rec (cache_dir,   0700);
    hc_mkdir     (session_dir, 0700);
  }
  else
  {
    profile_dir = install_dir;
    cache_dir   = install_dir;
    session_dir = install_dir;
    shared_dir  = install_dir;
  }
```

The current implementation is a problem as the `folder_config_init` function will later check for the OpenCL folder in the  `shared_dir`, and fail if its not found:

```c
  hc_asprintf (&cpath, "%s/OpenCL/", shared_dir);

  char *cpath_real = (char *) hcmalloc (PATH_MAX);

  if (realpath (cpath, cpath_real) == NULL)
  {
    event_log_error (hashcat_ctx, "%s: %s", cpath, strerror (errno));
...
    return -1;
  }
```

**Expected behavior**
Callers of the API should be able to control the folder config. 

**Hashcat version (please complete the following information):**
 - OS: Linux
 - Distribution: Debian 12
 - Version: Git HEAD

**Comment by ventaquil** [comment]
Had same problem here GH-2553.

I finally re-implemented PyHashcat on my own in Cython and got some working version with the trick below.

```py
cdef int session_init(hashcat_ctx_t *context):
    install_directory = HASHCAT_PREFIX
    shared_directory = f"{install_directory}/share/hashcat".encode()
    cdef char *shared_folder = shared_directory
    # todo describe why it is done in that way
    binary_path = os.readlink(f"/proc/{os.getpid()}/exe")
    binary_directory = os.path.dirname(binary_path).encode()
    cdef char *install_folder = binary_directory
    with nogil:
        return hashcat_session_init(context, install_folder, shared_folder, 0, NULL, 0)
```

The `HASHCAT_PREFIX` variable is set on build stage (by default `/usr/local` where my custom Hashcat lives).

I hope this can help.

**Comment by process0** [comment]
@ventaquil Thanks for adding your experience with this. I think it helps the case for adjusting the API or providing an additional init function where the folder config is respected.  I was able to get the library working the same way you described.  Its the same way pyhashcat does, and judging from the comments, id guess they were also confused by the behavior. 

```c
  char *py_path = "/usr/bin/";
  char *hc_path = "/usr/local/share/hashcat";
...
  /**
   *   !! IMPORTANT !!
   *   Getting the args to hashcat_session_init correct is critical.
   *   The first parameter is where Python is installed and
   *   the second is where you installed all the hashcat files.
   *
   * */
  self->rc_init = hashcat_session_init (self->hashcat_ctx, py_path, hc_path, 0, NULL, 0);
```

I don't think the work around constitutes a sane API.

**Comment by process0** [comment]
Actually,  pyhashcat's author did stumble on this in 2016...

https://github.com/hashcat/hashcat/issues/9#issuecomment-267238806

**Comment by ventaquil** [comment]
I fully agree that Hashcat should have a better alternative for the initialization process.

In my opinion Hashcat maintainers need to decide whether to make the API more user-friendly or consider removing this feature altogether.

Right now, creating a custom application with the Hashcat library involves a lot of workarounds and unconventional methods, usually figured out through challenging experimentation. The lack of comprehensive documentation only adds to the difficulty.

**Comment by process0** [comment]
@ventaquil If you can list other pain points you've encountered, that would be helpful. I plan on documenting issues I encounter.  I'm willing to help improve the API if this is something the maintainers are interested in.

**Comment by ventaquil** [comment]
@process0, the main issue I faced was related to versioning.

The binary and the library have the same version, tied to the binary. Between versions 6.2.2 and 6.2.3, if I recall correctly, there was a significant change in internal structures, necessitating recompilation after upgrading Hashcat.

Another significant issue is associated with the `user_options_t` structure. While this structure represents CLI input, it is used throughout the codebase.

Currently, I can set both `user_options->benchmark = true` and `user_options->hash_info = true` simultaneously. It would be more structured to have something like `hashcat_ctx->mode = ModeBenchmark`, where the `.mode` field is an enum type encompassing all possible modes. This would prevent the conflicting setting of benchmark and hash info at the same time.

The current situation leads to convoluted if-then-return logic, as illustrated in the following example:

https://github.com/hashcat/hashcat/blob/4d412c8e04dd972d1831aea8bead1a3b0cff3d8c/src/main.c#L195-L199

Finally, a crucial aspect is the data handling through the `hashcat_init` callback. To achieve proper event handling, the current approach involves global variables or unconventional methods, as there is only one global callback for all events. A more robust system would involve individual callbacks for each event. Also, to ensure the validity of this system in a multi-threading environment, all data should be passed as parameters to the respective callbacks instead of using `hashcat_ctx` to retrieve required data.

https://github.com/hashcat/hashcat/blob/4d412c8e04dd972d1831aea8bead1a3b0cff3d8c/src/main.c#L1160-L1221

In my PyHashcat binding, I've expanded the `hashcat_ctx_t` structure to effectively store events' callbacks.

```cython
typedef struct pyhashcat_ctx {
    struct hashcat_ctx;

    PyObject *events; // events dictionary
} pyhashcat_ctx_t;
```

The `events` attribute is essentially a simple `dict`, serving as a mapping between event IDs and their corresponding callback functions (which can be unset). This setup enables dynamic management of callbacks during runtime.
